rfcs/007 - Alternative Ways to Create Mappers.md

```diff
@@ -0,0 +1,122 @@
+
+We need to get the mapper-like behavior working. I'll explain some ways we might consider for that purpose.
+
+- Remote Mappers
+  - Mappers maintained inside IaSQL infrastructure
+  - Mappers maintained outside IaSQL infrastructure
+- In-IaSQL-Container Mappers
+  - Mapper Logic Inside IaSQL Engine
+  - Mapper Logic Inside Postgres
+
+### A Stand-Alone Mapper Design (Remote Mappers)
+
+We can run standalone versions of mappers. A standalone mapper means that its logic won't be inside the Postgres, nor in the engine code. But it'll run anywhere outside these two places. Therefore, we need a communication mechanism for this design (which can be something like gRPC, or HTTP as in other places). We also might want to run that standalone mapper on our own server, or on somewhere else. Two approaches:
+
+1. Standalone mappers are served on the IaSQL server: the mapper is submitted to our system, and we'll run it in a container. Then our engine will communicate with that mapper through a communication mechanism. This reduces latency, since the calls will be intra-network, but adds the burden of maintaining the standalone mappers to our side.
+![In-house Standalone Mapper](./assets/008/Mapper-1.png)
+2. Standalone mappers served on servers other than IaSQL: the mapper providers can host that mapper anywhere, and provide endpoints to the engine which will be called. This keeps the maintenance of the mapper on the mapper developer side, but increases the latency of mapper.
+![Self-hosted Standalone Mapper](./assets/008/Mapper-2.png)
+
+In the 2nd way, we should post the AWS credentials to the mapper through the internet, which should be done in a secure way. In both ways I don't see another possible approach than posting the AWS credentials to the mapper, since it needs them to query AWS APIs. Anyway, we should warn people to not use a mapper from untrusted authors, because the credentials will be shared with them.
+
+A positive point for having standalone mappers is that it's language agnostic: mapper developers can write the logic in whatever language they want, and it's fine, until they follow the guidelines for the Engine -> Mapper communication protocol.
+
+One negative point is that they'll be provided with the AWS credentials, and they can perform out-of-band data extraction on it on both of the above ways.
+
+### In-IaSQL-Container Mappers
+
+In this approach, the logic of the mappers will be executed on IaSQL: either on the Postgres server or on the Engine. Our current mappers are under this category.
+- Mapper logic is in IaSQL engine: `aws_acm`'s `CertificateMapper`, `aws_ecr`'s `RepositoryImageMapper`, etc.
+- Mapper logic is in Postgres server: This is where we initially called Pure-SQL module. Note that `aws_ecs_simplified` module can't be considered under this category, since it introduces no new mappers. It's playing in module-level (not mapper level) and handling a new Entity using already-existing mappers.
```

Comment:
Depending on how loosely you define a mapper, I would say that the function the triggers call are acting as mappers themselves. They are mapping an existing entity to another entity. However, my understand was that we were trying to get a JSON response from `aws_sdk` into a mapper.

Reply:
I'm defining a mapper to have these properties:
- A cloud id function
- An equality checker function
- An entity, that the mapper manages it
- Cloud CRUD functions
- Ability to call other mappers' CRUD functions

In that definition, I mean, the `ecs_simplified` does not fit since it doesn't provide a new mapper.

Reply:
Yes. It may be necessary to have a separate section on 3rd party modules vs 3rd party mappers, and also why 3rd party modules without 3rd party mappers was deemed insufficient. Specifically, `ecs_simplified` and other modules like it can only reduce down the functionality that has been already exposed through mappers we have written, it can't go beyond any functionality we didn't hand-expose, which makes us the bottleneck for the whole ecosystem.

Then it would be worth discussing why we need mapper-like behavior when the `aws_sdk` is now exposed directly. What do we lose if we just had a collection of triggers that directly call `aws_sdk` `invoke_` functions on INSERT/UPDATE/DELETE and execute in a cron job to do INSERT/UPDATE/DELETE calls into that table based on cloud changes.

(No eventual consistency, for one, but other issues arise, too. Would like it if you think about it. :) )